test/functional/eval/fnamemodify_spec.lua

```diff
@@ -36,4 +37,105 @@ describe('fnamemodify()', function()
   it(':8 works', function()
     eq('Xtest-fnamemodify.txt', fnamemodify([[Xtest-fnamemodify.txt]], ':8'))
   end)
+
+  it('handles examples from ":help filename-modifiers"', function()
+    local filename = "src/version.c"
+    local cwd = getcwd()
+
+    eq(cwd .. '/src/version.c', fnamemodify(filename, ':p'))
+    eq('src/version.c', fnamemodify(filename, ':p:.'))
+    eq(cwd .. '/src', fnamemodify(filename, ':p:h'))
+    eq(cwd .. '', fnamemodify(filename, ':p:h:h'))
+    eq('version.c', fnamemodify(filename, ':p:t'))
+    eq(cwd .. '/src/version', fnamemodify(filename, ':p:r'))
+
+    eq(cwd .. '/src/main.c', fnamemodify(filename, ':s?version?main?:p'))
+
+    local converted_cwd = cwd:gsub('/', '\\')
+    eq(converted_cwd .. '\\src\\version.c', fnamemodify(filename, ':p:gs?/?\\\\?'))
+
+    eq('src', fnamemodify(filename, ':h'))
+    eq('version.c', fnamemodify(filename, ':t'))
+    eq('src/version', fnamemodify(filename, ':r'))
+    eq('version', fnamemodify(filename, ':t:r'))
+    eq('c', fnamemodify(filename, ':e'))
+
+    eq('src/main.c', fnamemodify(filename, ':s?version?main?'))
+  end)
+
+  it('handles advanced examples from ":help filename-modifiers"', function()
+    local filename = "src/version.c.gz"
+
+    eq('gz', fnamemodify(filename, ':e'))
+    eq('c.gz', fnamemodify(filename, ':e:e'))
+    eq('c.gz', fnamemodify(filename, ':e:e:e'))
+
+    eq('c', fnamemodify(filename, ':e:e:r'))
+
+    eq('src/version.c', fnamemodify(filename, ':r'))
+    eq('c', fnamemodify(filename, ':r:e'))
+
+    eq('src/version', fnamemodify(filename, ':r:r'))
+    eq('src/version', fnamemodify(filename, ':r:r:r'))
+  end)
+
+  it('handles :h', function()
+    eq('.', fnamemodify('hello.txt', ':h'))
+
+    eq('path/to', fnamemodify('path/to/hello.txt', ':h'))
+  end)
+
+  it('handles :t', function()
+    eq('hello.txt', fnamemodify('hello.txt', ':t'))
+    eq('hello.txt', fnamemodify('path/to/hello.txt', ':t'))
+  end)
+
+  it('handles :r', function()
+    eq('hello', fnamemodify('hello.txt', ':r'))
+    eq('path/to/hello', fnamemodify('path/to/hello.txt', ':r'))
+  end)
+
+  it('handles :e', function()
+    eq('txt', fnamemodify('hello.txt', ':e'))
+    eq('txt', fnamemodify('path/to/hello.txt', ':e'))
+  end)
+
+  it('handles regex replacements', function()
+    eq('content-there-here.txt', fnamemodify('content-here-here.txt', ':s/here/there/'))
+    eq('content-there-there.txt', fnamemodify('content-here-here.txt', ':gs/here/there/'))
+  end)
+
+  it('handles shell escape', function()
+    eq("'hello there! quote '\\'' newline\n'", fnamemodify("hello there! quote ' newline\n", ':S'))
```

Comment:
AppVeyor (Windows) CI is failing, maybe because `\r` is added.  Untested suggestion:

```suggestion
    eq(string.format("'hello there! quote '\\'' newline%s'", (iswin() and '\r\n' or '\n')),
      fnamemodify("hello there! quote ' newline\n", ':S'))
```

Reply:
Looks like it's differences in how we expand for `cmd.exe` as opposed to a posix shell - I've gone back to your `[[ ... ]]` suggestion too, I think it now helps make reading the differences between the Windows and non-Windows versions (hopefully) clear.